File: src/game/hardwareinput/MouseControllerManager.java
package game.hardwareinput;

import game.Bomberman;
import game.tasks.PeriodicTask;
import game.entity.Player;
import game.entity.models.Entity;
import game.entity.models.Coordinates;
import game.entity.models.Direction;

import java.awt.event.*;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * The MouseControllerManager class handles mouse interactions and controls player movement based on mouse input.
 */
public class MouseControllerManager extends MouseAdapter implements MouseMotionListener {
    private boolean mouseDraggedInteractionOccured;
    private boolean mouseDraggedInteractionInterrupted;
    private boolean mouseDragInteractionEntered;
    private boolean isMouseDragged;
    private boolean isMouseClicked;
    private LinkedList<MouseEvent> mouseClickQueue = new LinkedList<>();
    // First and last directions from the player are differentiated so the task can be terminated when they have no direction in common,
    // in order to avoid an infinite loop
    private List<Direction> firstDirectionsFromPlayer = new ArrayList<>();
    private List<Direction> latestDirectionsFromPlayer = new ArrayList<>();
    private Coordinates mouseCoords;
    private final int DELAY = 300;
    private Entity entity = null;

    private final Runnable task = () -> {
        Player currPlayer = Bomberman.getMatch().getPlayer();
        if(currPlayer == null)return;
        onCooldown();

        //interact with mouse
        if (Bomberman.getMatch().getPlayer() != null && !currPlayer.getAliveState())
            return;

        if (entity != null && Bomberman.getMatch().getPlayer().getListClassInteractWithMouseClick().contains(entity.getClass())) {
            entity.mouseInteractions();
            onPeriodicTaskEnd();
            nextCommandInQueue();
            return;
        }

        latestDirectionsFromPlayer = mouseCoords.fromCoordinatesToDirection(currPlayer.getCoords());
        for (Direction d : latestDirectionsFromPlayer) {
            Bomberman.getMatch().getControllerManager().onKeyPressed(d.toCommand());
        }

        if (firstDirectionsFromPlayer.isEmpty() || firstDirectionsFromPlayer.stream().anyMatch(d -> latestDirectionsFromPlayer.contains(d))) {
            return;
        }

        onPeriodicTaskEnd();
        nextCommandInQueue();
    };

    /**
     * Retrieves and removes the next command in the mouseClickQueue.
     */
    public void nextCommandInQueue(){
        mouseClickQueue.poll();
        if(!mouseClickQueue.isEmpty())
        onMouseClicked(mouseClickQueue.peekFirst());
    }

    /**
     * Called when the mouse button is released.
     *
     * @param event The MouseEvent representing the release of the mouse button.
     */
    @Override
    public void mouseReleased(MouseEvent event){
        // If a mouse dragged interaction was not successful and the mouse is released, call mouseClicked
        // so that click-dependent interactions can be called.
        // This is necessary because mouseClicked is never called when mouseDragged is called, unlike mousePressed.
        // mouseClicked was still chosen because the two methods shouldn't be called at exactly the same time.
        if (!isMouseDragged) {
            return;
        }
        if (!mouseDraggedInteractionOccured) mouseClicked(event);
        mouseDragInteractionEntered = false;
        isMouseDragged = false;
        mouseDraggedInteractionOccured = false;
        mouseDraggedInteractionInterrupted = false;
    }

    /**
     * Called when the mouse button is pressed.
     *
     * @param event The MouseEvent representing the press of the mouse button.
     */
    @Override
    public void mousePressed(MouseEvent event){
        if(isMouseClicked) return;
        mouseCoords = new Coordinates(event.getX(), event.getY());
        entity = Coordinates.getEntityOnCoordinates(mouseCoords);
    }

    private final PeriodicTask playerMovementTask = new PeriodicTask(task, DELAY);

    @Override
    public void mouseClicked(MouseEvent e) {
        if(!mouseClickQueue.isEmpty()) {
            mouseClickQueue.add(e);
            return;
        }

        mouseClickQueue.add(e);
        onMouseClicked(e);
    }

    public void onMouseClicked(MouseEvent e){
        if(isMouseClicked) return;
        isMouseClicked = true;

        Player player = Bomberman.getMatch().getPlayer();
        if(player == null || !player.getAliveState()) {
            return;
        }

        //isMouseClicked = true;
        mouseCoords = new Coordinates(e.getX(), e.getY());
        firstDirectionsFromPlayer = mouseCoords.fromCoordinatesToDirection(Bomberman.getMatch().getPlayer().getCoords());
        entity = Coordinates.getEntityOnCoordinates(mouseCoords);
        playerMovementTask.resume();
    }

    // Directions are refreshed and will be replaced in task
    public void onCooldown() {
        for (Direction d: firstDirectionsFromPlayer) {
            Bomberman.getMatch().getControllerManager().onKeyReleased(d.toCommand());
        }
    }

    public Coordinates getMouseCoords() {
        return mouseCoords;
    }

    public void setMouseDraggedInteractionOccured(boolean mouseDraggedInteractionOccured) {
        this.mouseDraggedInteractionOccured = mouseDraggedInteractionOccured;
    }

    public boolean isMouseDraggedInteractionInterrupted() {
        return mouseDraggedInteractionInterrupted;
    }

    public void setMouseDraggedInteractionInterrupted(boolean mouseDraggedInteractionInterrupted) {
        this.mouseDraggedInteractionInterrupted = mouseDraggedInteractionInterrupted;
    }

    public boolean isMouseDragInteractionEntered() {
        return mouseDragInteractionEntered;
    }

    public void setMouseDragInteractionEntered(boolean mouseDragInteractionEntered) {
        this.mouseDragInteractionEntered = mouseDragInteractionEntered;
    }

    public boolean isMouseDragged() {
        return isMouseDragged;
    }

    public boolean isMouseClicked() {
        return isMouseClicked;
    }

    public void stopPeriodicTask(){
        mouseClickQueue = new LinkedList<>();
        onPeriodicTaskEnd();
    }

    public void onPeriodicTaskEnd(){
        if(!isMouseClicked)
            return;

        playerMovementTask.stop();
        onCooldown();
        isMouseClicked = false;
    }

    @Override
    public void mouseDragged(MouseEvent event){
        if(isMouseClicked) return;

        isMouseDragged = true;
        mouseCoords = new Coordinates(event.getX(), event.getY());
        if (entity == null) return;
        entity.mouseInteractions();
    }

    public Entity getEntity(){
        return entity;
    }

}

File: src/game/hardwareinput/MouseControllerManager.kt
package game.hardwareinput

import game.Bomberman
import game.entity.models.Coordinates
import game.entity.models.Direction
import game.entity.models.Entity
import game.tasks.PeriodicTask
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionListener
import java.util.*

/**
 * The MouseControllerManager class handles mouse interactions and controls player movement based on mouse input.
 */
class MouseControllerManager : MouseAdapter(), MouseMotionListener {
    private val DELAY = 300
    private var mouseDraggedInteractionOccured = false
    var isMouseDraggedInteractionInterrupted = false
    var isMouseDragInteractionEntered = false
    var isMouseDragged = false
        private set
    var isMouseClicked = false
        private set
    private var mouseClickQueue = LinkedList<MouseEvent>()

    // First and last directions from the player are differentiated so the task can be terminated when they have no direction in common,
    // in order to avoid an infinite loop
    private var firstDirectionsFromPlayer: List<Direction> = ArrayList()
    private var latestDirectionsFromPlayer: List<Direction> = ArrayList()
    var mouseCoords: Coordinates? = null
        private set
    var entity: Entity? = null
        private set

    /**
     * Retrieves and removes the next command in the mouseClickQueue.
     */
    private fun nextCommandInQueue() {
        mouseClickQueue.poll()

        if (!mouseClickQueue.isEmpty())
            onMouseClicked(mouseClickQueue.peekFirst())
    }

    private val movementTask = Runnable {
        val match = Bomberman.getMatch()
        val currPlayer = match.player ?: return@Runnable
        onCooldown()

        if (!currPlayer.aliveState) return@Runnable

        val entityToInteract = entity
        if (entityToInteract != null && match.player.listClassInteractWithMouseClick.contains(entityToInteract.javaClass)) {
            entityToInteract.mouseInteractions()
            onMovementPeriodicTaskEnd()
            nextCommandInQueue()
            return@Runnable
        }

        latestDirectionsFromPlayer = mouseCoords?.fromCoordinatesToDirection(currPlayer.coords) ?: emptyList()

        latestDirectionsFromPlayer.forEach {
            match.controllerManager.onKeyPressed(it.toCommand())
        }

        if (firstDirectionsFromPlayer.isNotEmpty() && firstDirectionsFromPlayer.none { it !in latestDirectionsFromPlayer }) {
            return@Runnable
        }
        onMovementPeriodicTaskEnd()
        nextCommandInQueue()
    }

    /**
     * Called when the mouse button is released.
     *
     * @param event The MouseEvent representing the release of the mouse button.
     */
    override fun mouseReleased(event: MouseEvent) {
        // If a mouse dragged interaction was not successful and the mouse is released, call mouseClicked
        // so that click-dependent interactions can be called.
        // This is necessary because mouseClicked is never called when mouseDragged is called, unlike mousePressed.
        // mouseClicked was still chosen because the two methods shouldn't be called at exactly the same time.
        if (!isMouseDragged) {
            return
        }
        if (!mouseDraggedInteractionOccured) mouseClicked(event)
        isMouseDragInteractionEntered = false
        isMouseDragged = false
        mouseDraggedInteractionOccured = false
        isMouseDraggedInteractionInterrupted = false
    }

    /**
     * Called when the mouse button is pressed.
     *
     * @param event The MouseEvent representing the press of the mouse button.
     */
    override fun mousePressed(event: MouseEvent) {
        if (isMouseClicked) return
        mouseCoords = Coordinates(event.x, event.y)
        entity = Coordinates.getEntityOnCoordinates(mouseCoords)
    }

    override fun mouseClicked(e: MouseEvent) {
        if (!mouseClickQueue.isEmpty()) {
            mouseClickQueue.add(e)
            return
        }
        mouseClickQueue.add(e)
        onMouseClicked(e)
    }

    fun onMouseClicked(e: MouseEvent) {
        if (isMouseClicked) return
        isMouseClicked = true
        val player = Bomberman.getMatch().player
        if (player == null || !player.aliveState) {
            return
        }

        //isMouseClicked = true;
        mouseCoords = Coordinates(e.x, e.y)
        firstDirectionsFromPlayer = mouseCoords!!.fromCoordinatesToDirection(Bomberman.getMatch().player.coords)
        entity = Coordinates.getEntityOnCoordinates(mouseCoords)
        startMouseTasks()
    }


    private val playerMovementTask = PeriodicTask(movementTask, DELAY)

    // Directions are refreshed and will be replaced in task
    fun onCooldown() {
        for (d in firstDirectionsFromPlayer) {
            Bomberman.getMatch().controllerManager.onKeyReleased(d.toCommand())
        }
    }

    fun setMouseDraggedInteractionOccured(mouseDraggedInteractionOccured: Boolean) {
        this.mouseDraggedInteractionOccured = mouseDraggedInteractionOccured
    }

    fun stopMovementTask() {
        mouseClickQueue = LinkedList()
        onMovementPeriodicTaskEnd()
    }

    private fun onMovementPeriodicTaskEnd() {
        if (!isMouseClicked) return
        playerMovementTask.stop()
        onCooldown()
        isMouseClicked = false
    }

    override fun mouseDragged(event: MouseEvent) {
        if (isMouseClicked) return
        isMouseDragged = true
        mouseCoords = Coordinates(event.x, event.y)
        if (entity == null) return
        entity!!.mouseInteractions()
    }
    fun startMouseTasks(){
        playerMovementTask.resume()
    }
}